apps/speed.c

```diff
@@ -3548,13 +3548,14 @@ int speed_main(int argc, char **argv)
             size_t send_secret_len, out_len;
             size_t rcv_secret_len;
             unsigned char *out = NULL, *send_secret = NULL, *rcv_secret;
-            size_t bits;
+            unsigned int bits;
             char *name;
+            char sfx[100];
             OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END };
             int use_params = 0;
             enum kem_type_t { KEM_RSA = 1, KEM_EC, KEM_X25519, KEM_X448 } kem_type;
 
-            if (strncmp(kem_name, "rsa", 3) == 0)
+            if (sscanf(kem_name, "rsa%u%s", &bits, sfx) == 1)
```

Comment:
Sure: Without this change, any algorithm (name) beginning with "rsa..." would lead to this (KEM&SIG) algorithm being considered as a "classic" RSA variant, leading to them being configured as such (e.g., [here for the bit strength](https://github.com/openssl/openssl/blob/6aeb42eca97227c8235af0986d1525ee4a916504/apps/speed.c#L3575-L3576)). This however is incorrect, if hybrid (classic combined with PQ) algorithms' names are similarly formed, e.g., "rsa3072_dilithium2". Thus, this PR restricts the currently (due to my own mistake in #19968) overly lax interpretation of "anything starting with 'rsa' _is_ RSA" to "only algorithm names with exactly the letters 'rsa' followed by a digit and nothing else is RSA".

Reply:
Ah...its the "and nothing else" bit which I was not seeing (i.e. implemented via checking the return value of `sscanf` is exactly 1). Perhaps a comment here might have helped?

Reply:
Agreed. Added in https://github.com/openssl/openssl/pull/20898/commits/d7725875f374bacae548fbf7674ce0ae07273248.

Reply:
Ah, so `sfx` is for `suffix`!

What happens if a user supplies a string with such a suffix that's > 100 characters? Won't it overflow the `sfx[]` on the stack? Yes, we'll then return with an error, but on a system without stack protection that could be bad.

How about checking the length of the string before trying to parse it into a fixed-length buffer?

Reply:
Agreed. Added in https://github.com/openssl/openssl/pull/20898/commits/6f3c26321c73862e472f32851c5b00dabf815b71.